dulwich/porcelain.py

```diff
@@ -1859,6 +1858,95 @@ def reset_file(repo, file_path: str, target: bytes = b'HEAD',
+      target_tree: A gut object for the target tree state
+    Returns: List of tracked changed files
+    """
+    tracked_changes = []
+    for change in repo.open_index().changes_from_tree(repo.object_store, target_tree.id):
+        file = change[0][0] or change[0][1]  # No matter whether the file is added, modified, or deleted.
+        try:
+            current_entry = current_tree.lookup_path(repo.object_store.__getitem__, file)
+        except KeyError:
+            current_entry = None
+        try:
+            target_entry = target_tree.lookup_path(repo.object_store.__getitem__, file)
+        except KeyError:
+            target_entry = None
+
+        if current_entry or target_entry:
+            tracked_changes.append(file.decode())
+
+    return tracked_changes
+
+
+def checkout(repo, target: bytes, force: bool = False):
+    """switch branches or restore working tree files
+    Args:
+      repo: dulwich Repo object
+      target: branch name or commit sha to checkout
+      force: true or not to force checkout
+    """
+    # Check repo status.
+    if not force:
+        index = repo.open_index()
+        for file in get_tree_changes(repo)['modify']:
+            if file in index:
+                raise DirNotCleanError('trying to checkout when working directory not clean')
+
+        normalizer = repo.get_blob_normalizer()
+        filter_callback = normalizer.checkin_normalize
+
+        unstaged_changes = list(get_unstaged_changes(index, repo.path, filter_callback))
+        for file in unstaged_changes:
+            if file in index:
+                raise DirNotCleanError('trying to checkout when working directory not clean')
+
+    current_tree = parse_tree(repo, repo.head())
+    target_tree = parse_tree(repo, target)
+
+    # Update head.
+    if target == b'HEAD':  # Do not update head while trying to checkout to HEAD.
+        pass
+    elif target in repo.refs.keys(base=LOCAL_BRANCH_PREFIX):
+        update_head(repo, target)
+    else:
+        # If checking out a remote branch, create a local one without the remote name prefix.
+        config = repo.get_config()
+        name = target.split(b"/")[0]
+        section = (b"remote", name)
+        if config.has_section(section):
+            checkout_target = target.replace(name + b"/", b"")
+            try:
+                branch_create(repo, checkout_target, (LOCAL_REMOTE_PREFIX + target).decode())
+            except Error:
+                pass
+            update_head(repo, LOCAL_BRANCH_PREFIX + checkout_target)
+            target_tree = parse_tree(repo, checkout_target)
+        else:
+            update_head(repo, target, detached=True)
+
+    # Un-stage files in the current_tree or target_tree.
+    tracked_changes = _determine_tracked_changes(repo, current_tree, target_tree)
+    repo.unstage(tracked_changes)
+
+    repo.reset_index(target_tree.id)
+
+    # Remove the untracked file which are in the current_file_set.
```

Comment:
Wouldn't it be better to get the list of files in the previous index and diff that against the new index?

Reply:
Possibly, I am not super familiar with this.  Any hints or existing code that does this that I can look at?

Reply:
calling repo.open_index() will get you the current index (before making changes to it with repo.unse or repo.reset_index, obviously)

the index object provides an iterator that returns all the relative paths in the index, as bytestrings.